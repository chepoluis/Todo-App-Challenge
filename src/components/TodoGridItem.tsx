import { formatDate } from '../helpers/commons';
import { getTodos } from '../helpers/getTodos';
import { colors } from '../helpers/colors';

import './styles/TodoGridItem.css';

export const TodoGridItem = ( props: any ) => {
    const { description, dueDate, isComplete, id } = props;

    const complete = () => {
        getTodos('PATCH', id).then(({ status }) => {
            if (status === 'success') {
                props.setTodos((val: any) => { 
                    return val.map((todo: any) => {
                        
                        if (id === todo.id) {
                            return { ...todo, isComplete: !todo.isComplete }
                        }

                        return todo;
                    })
                })
            }
        })
    }
    
    const overdue = () => {
        const today = new Date();
        const todoDate = new Date(dueDate);

        return todoDate.getTime() > today.getTime();
    }

    return (
        <div className = "card-wrapper mr-5" style={{ "marginRight": "20px", "marginBottom": "15px" }}>
            <div className = "card-top" style={{"backgroundColor": colors[1].primaryColor}}></div>
            <div className = "task-holder">
                <span className = "card-header" style={{"backgroundColor": isComplete && colors[4].primaryColor, "borderRadius": "10px"}}>{ isComplete ? 'Completed' : 'Not completed' }</span>
                <p className = "mt-3">{description}</p>
                <p>{ formatDate(dueDate) }</p>
                { 
                    (!overdue() && !isComplete && dueDate != null) && <p>Overdue</p>
                }

                <div style={{"position": "absolute", "right" : "20px", "bottom" : "20px"}}>
                    <i className = "far fa-ballot-check mr-3" style={{"color" : colors[4].primaryColor, "cursor" : "pointer"}} onClick={ complete }></i>
                </div>
            </div>
        </div>
    );
};
