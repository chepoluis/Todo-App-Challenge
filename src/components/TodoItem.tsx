import { useFetch } from '../hooks/useFetch';

export const TodoItem = () => {
    const { loading, data } = useFetch('https://944ba3c5-94c3-4369-a9e6-a509d65912e2.mock.pstmn.io/get', 'GET');

    console.log(data)

    return (
        <div>
            <h1>Todo Item</h1>

            {
                loading ?
                    <div className='alert alert-info text-center'>
                        Loading...
                    </div>
                :
                    <h1>:D</h1>
            }
        </div>
    );
};
