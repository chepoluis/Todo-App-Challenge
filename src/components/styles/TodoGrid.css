.header{
    height: 100px;
    width: 100%;
    background-color : #b2c8eb;
    padding-top: 30px;
}

.todo-grid {
    height: 600px;
    width: 100%;
    display: flex;
    flex-wrap: wrap;
    padding : 40px 100px;
}
