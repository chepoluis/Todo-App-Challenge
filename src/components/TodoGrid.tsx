import { TodoGridItem } from "./TodoGridItem";

import { useEffect, useState } from "react";
import { getTodos } from '../helpers/getTodos';

import './styles/TodoGrid.css';

export const TodoGrid = () => {
    const [todos, setTodos] = useState([]);
    const [loading, setLoading] = useState(true);

    useEffect(() => {
        setLoading(true);

        getTodos('GET').then((todos) => {
            setTodos(todos);

            setLoading(false);
        })
    }, []);
    
    const sortByDueDate = ( ) => {
        const copy = [...todos];
        copy.sort((a: any, b: any) => {
            const aDate: any = new Date(a.dueDate);
            const bDate: any = new Date(b.dueDate);

            return aDate - bDate;
        })

        setTodos(copy);
    }

    return (
        <>
            <div className="header text-center">
                <h3>Todo App</h3>
            </div>
            {
                loading ?
                    <div className='alert alert-info text-center'>
                        Loading...
                    </div>
                :
                <div>
                    <button className="btn btn-outline-primary" style={{ "marginTop": "10px", "marginLeft": "10px" }} onClick={ sortByDueDate }>
                        Order by due date
                    </button>

                    <div className="todo-grid">
                        {
                            todos.map((t: any) => {
                                return <TodoGridItem key={ t.id } {...t} setTodos={ setTodos } />
                            })
                        }
                    </div>
                </div>
            }
        </>
    );
};
