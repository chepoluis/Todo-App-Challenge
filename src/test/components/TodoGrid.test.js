import { shallow } from "enzyme";
import { TodoGrid } from "../../components/TodoGrid";

describe('Test in <TodoGrid /> component', () => {
    
    test('should show the component', () => {
        const wrapper = shallow( <TodoGrid />);
        expect( wrapper ).toMatchSnapshot();
    });
    
    test('should show Loading... if there is no data yet', () => {
        const wrapper = shallow( <TodoGrid />);
        const message = 'Loading...';

        expect( wrapper.find('.alert').props().children ).toBe( message );
    });
})