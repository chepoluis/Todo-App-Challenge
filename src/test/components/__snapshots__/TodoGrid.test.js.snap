// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Test in <TodoGrid /> component should show the component 1`] = `
<Fragment>
  <div
    className="header text-center"
  >
    <h3>
      Todo App
    </h3>
  </div>
  <div
    className="alert alert-info text-center"
  >
    Loading...
  </div>
</Fragment>
`;
